### PR TITLE
ci: s3 cache

### PR DESCRIPTION
**Pull request by shumkov**
## Issue being fixed or feature implemented

Self-hosted runners are expensive so we need to utilize cache as much as possible to reduce costs.
This PR is using s3 store as a cache backend for NPM, Cargo, Docker build and Rust build caches in CI

There is some work needed to support ARM builds in a proper way

## What was done?

- Disabled sccache by default
- Enabled docker mount cache
- Enabled docker layer cache
- Use independent envs to setup caching for each service in Dashmate
- Utilized s3 for the docker layer and mount caches in CI

This PR contains prelumeary work from https://github.com/dashpay/platform/pull/1105 and https://github.com/dashpay/platform/pull/953

## How Has This Been Tested?

Running CI and locally

## Breaking Changes

None

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have made corresponding changes to the documentation

**For repository code-owners and collaborators only**
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

File: .github/workflows/rs-drive-abci.yml [hunk 1]
```diff
@@ -33,15 +33,16 @@ jobs:
...
-    timeout-minutes: 180  # ARM build can take forever
-    steps:
-      - uses: actions/checkout@v3
-      - uses: ./.github/actions/docker
-        with:
-          dockerfile: ./Dockerfile
-          image: dashpay/drive-abci
-          target: drive-abci
-          push: false
+# TODO: Remove?
```
Comment (strophy) [inline .github/workflows/rs-drive-abci.yml:36]: What was the purpose of this? Just to check if Docker build works?

**Review by strophy — CHANGES_REQUESTED** [review]
This is looking really good! Let's make a few more changes for readability

File: .github/actions/docker/action.yaml [hunk 2]
```diff
@@ -2,48 +2,86 @@
...
+      uses: docker/login-action@v2
+      with:
+        username: ${{ inputs.dockerhub_username }}
+        password: ${{ inputs.dockerhub_token }}
+      if: ${{ inputs.dockerhub_token != '' }}
+
+    - name: Configure AWS credentials
+      uses: aws-actions/configure-aws-credentials@v2
+      with:
+        aws-region: eu-west-1
```
[the remark this answers is not in the document]
Comment (shumkov) [reply .github/actions/docker/action.yaml:45]: I'm not sure. Do s5cmd and buildkit use their own authorization approach?

File: .github/workflows/all-packages.yml [hunk 3]
```diff
@@ -81,73 +77,69 @@ jobs:
...
+          CACHE_DASHMATE_HELPER_TO: type=s3,region=eu-west-1,bucket=platform-runner-cache,prefix=cache-layers/,name=dashpay/dashmate-helper_${{ github.sha }};dashpay/dashmate-helper_${{ github.head_ref }};dashpay/dashmate-helper,mode=max
+          CACHE_ENVOY_FROM: type=s3,region=eu-west-1,bucket=platform-runner-cache,prefix=cache-layers/,name=dashpay/envoy_${{ github.sha }};dashpay/envoy_${{ github.head_ref }};dashpay/envoy
+          CACHE_ENVOY_TO: type=s3,region=eu-west-1,bucket=platform-runner-cache,prefix=cache-layers/,name=dashpay/envoy_${{ github.sha }};dashpay/envoy_${{ github.head_ref }};dashpay/envoy,mode=max
+          CACHE_DEPS_FROM: type=s3,region=eu-west-1,bucket=platform-runner-cache,prefix=cache-layers/,name=dashpay/deps_${{ github.sha }};dashpay/deps_${{ github.head_ref }};dashpay/deps
+          CACHE_DEPS_TO: type=s3,region=eu-west-1,bucket=platform-runner-cache,prefix=cache-layers/,name=dashpay/deps_${{ github.sha }};dashpay/deps_${{ github.head_ref }};dashpay/deps,mode=max
```
[the remark this answers is not in the document]
Comment (shumkov) [reply .github/workflows/all-packages.yml:131]: Ok, I will figure out something )

File: .github/workflows/rs-drive-abci.yml (hunk 1 above)
Comment (shumkov) [reply .github/workflows/rs-drive-abci.yml:36]: Yes. There is no need for that since we build Drive in test suite workflow.

**Review by strophy — APPROVED** [review]
:rocket:

**Pull request merged**